File: Problem21.java
// Time Complexity : O(N)
// Space Complexity : O(2N) 
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this: understood
// Summary: //In this problem we are creating a 2 column matrix of taking or skipping the house
            //skip condition takes max from row above
            //take condition adds the value in array to the skip of row above

class Solution {
    public int rob(int[] nums) {
        if(nums == null || nums.length == 0) return 0;
        int skip = 0;
        int take = nums[0];
        for(int i = 1; i < nums.length; i++){
            int temp = skip;
            //0
            skip = Math.max(skip,take);
            //1
            take = temp + nums[i];
            ///System.out.println("Skip: " + skip + " " + "Take: " + take);
        }
        return Math.max(skip, take);
    }
}

File: Problem_21.py
# Time Complexity: O(n) where  n => houses
# Space Complexity: O(1)

class Solution:
    def rob(self, nums: List[int]) -> int:
        if len(nums) <= 1:
            return nums[0]
        dp = [0] * len(nums)
        prev = nums[0]
        curr = max(prev, nums[1]+0)

        for i in range(2, len(nums)):
            temp = curr
            curr = max(curr,  nums[i] + prev)
            prev = temp

        return curr